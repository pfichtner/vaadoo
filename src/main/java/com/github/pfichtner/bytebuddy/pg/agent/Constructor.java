package com.github.pfichtner.bytebuddy.pg.agent;

import static lombok.AccessLevel.PRIVATE;

import java.util.Set;

import javax.validation.ConstraintViolation;
import javax.validation.ConstraintViolationException;
import javax.validation.Validation;

import lombok.NoArgsConstructor;
import net.bytebuddy.asm.Advice;

@NoArgsConstructor(access = PRIVATE)
public final class Constructor {

	@Advice.OnMethodExit
	public static void exitConstructor(@Advice.This Object thisObj) {
		Set<ConstraintViolation<Object>> validationResult = Validation.buildDefaultValidatorFactory().getValidator()
				.validate(thisObj);
		if (!validationResult.isEmpty()) {
			throw new ConstraintViolationException(validationResult);
		}
	}

}