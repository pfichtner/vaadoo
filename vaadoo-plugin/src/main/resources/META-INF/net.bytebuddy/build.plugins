com.github.pfichtner.vaadoo.AddJsr380ValidationPlugin
