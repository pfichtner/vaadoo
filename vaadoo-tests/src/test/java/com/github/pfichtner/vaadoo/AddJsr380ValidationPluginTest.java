package com.github.pfichtner.vaadoo;

import static java.lang.ClassLoader.getSystemClassLoader;
import static java.util.Collections.singletonMap;
import static org.approvaltests.Approvals.verify;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.stream.Stream;

import org.jetbrains.java.decompiler.main.decompiler.ConsoleDecompiler;
import org.jetbrains.java.decompiler.main.extern.IFernflowerPreferences;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import com.example.SomeClass;
import com.example.SomeLombokClass;
import com.example.SomeRecord;

import net.bytebuddy.ByteBuddy;
import net.bytebuddy.build.Plugin;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.loading.ByteArrayClassLoader;

class AddJsr380ValidationPluginTest {

	@Test
	void testGeneratedBytecode(@TempDir Path tempDir) throws Exception {
		verify(decompile(tempDir, SomeClass.class));
	}

	@Test
	void testGeneratedBytecodeOnLombokClass(@TempDir Path tempDir) throws Exception {
		verify(decompile(tempDir, SomeLombokClass.class));
	}

	@Test
	void testGeneratedBytecodeOnRecord() throws Exception {
		try (AddJsr380ValidationPlugin sut = new AddJsr380ValidationPlugin()) {
			var constructor = firstConstructor(transform(sut, SomeRecord.class));
			assertThatExceptionOfType(InvocationTargetException.class).isThrownBy(() -> {
				constructor.newInstance(null, sut, null, null, null, null, null, null, null, "", false, false, null,
						null, 0, Long.valueOf(42), Short.valueOf((short) 42));
			}).satisfies(e -> assertThat(e.getCause()).isInstanceOf(NullPointerException.class)
					.hasMessageContaining("someNotEmptyCharSequence must not be null"));
		}
	}

	private String decompile(Path tempDir, Class<?> clazz) throws URISyntaxException, IOException {
		Path sourcePath = Path.of(getClass().getResource("/" + clazz.getName().replace('.', '/') + ".class").toURI());
		Files.copy(sourcePath, tempDir.resolve(clazz.getSimpleName() + ".class"));
		return decompileClass(tempDir, clazz);
//		var transformedClass = transform(sut, SomeClass.class);
//		firstConstructor(transformedClass).newInstance("", "", "", "", false);
	}

	private static Constructor<?> firstConstructor(Class<?> clazz) {
		return Stream.of(clazz.getConstructors()).findFirst().get();
	}

	private static Class<?> transform(Plugin plugin, Class<?> clazz) throws ClassNotFoundException {
		var builder = new ByteBuddy().redefine(clazz);
		var transformed = plugin.apply(builder, TypeDescription.ForLoadedType.of(clazz), null).make();
		var classname = clazz.getName();
		var classLoader = new ByteArrayClassLoader(getSystemClassLoader(),
				singletonMap(classname, transformed.getBytes()));
		return classLoader.loadClass(classname);
	}

	private String decompileClass(Path destination, Class<?> clazz) throws IOException {
		Map<String, Object> options = Map.of( //
				IFernflowerPreferences.REMOVE_BRIDGE, "true", //
				IFernflowerPreferences.DECOMPILE_GENERIC_SIGNATURES, "true" //
		);
		ConsoleDecompiler decompiler = new ConsoleDecompiler(destination.toFile(), options);
		decompiler.addSpace(destination.toFile(), true);
		decompiler.decompileContext();

		Path decompiledJavaFile = destination.resolve(clazz.getSimpleName() + ".java");
		String decompiledSource = Files.readString(decompiledJavaFile);
		Files.delete(decompiledJavaFile);
		return decompiledSource;
	}

}
