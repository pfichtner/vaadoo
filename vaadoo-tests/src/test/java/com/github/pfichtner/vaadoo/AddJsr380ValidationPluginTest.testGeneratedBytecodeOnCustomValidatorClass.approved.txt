.bytecode 61.0
.class public synchronized com/example/custom/ClassWithFizzNumber
.super java/lang/Object
.inner class public static final Lookup inner java/lang/invoke/MethodHandles$Lookup outer java/lang/invoke/MethodHandles

.field private final 'number' Ljava/lang/Integer;

.method public <init>(Ljava/lang/Integer;)V
aload 0
aload 1
invokestatic com/example/custom/ClassWithFizzNumber/validate(Ljava/lang/Integer;)V
invokespecial java/lang/Object/<init>()V
aload 0
aload 1
putfield com/example/custom/ClassWithFizzNumber/number Ljava/lang/Integer;
return
.limit locals 2
.limit stack 2
.end method
.dep com/example/custom/ClassWithFizzNumber;validate
.dep java/lang/Object;<init>
.dep com/example/custom/ClassWithFizzNumber;number
.provide com/example/custom/ClassWithFizzNumber;<init>
.desc Ljava/lang/Integer;
.desc ()V
.desc (Ljava/lang/Integer;)V

.method public <init>(Ljava/lang/Integer;Z)V
aload 0
aload 1
iload 2
invokestatic com/example/custom/ClassWithFizzNumber/validate(Ljava/lang/Integer;Z)V
invokespecial java/lang/Object/<init>()V
aload 0
aload 1
putfield com/example/custom/ClassWithFizzNumber/number Ljava/lang/Integer;
return
.limit locals 3
.limit stack 3
.end method
.dep com/example/custom/ClassWithFizzNumber;validate
.dep java/lang/Object;<init>
.dep com/example/custom/ClassWithFizzNumber;number
.provide com/example/custom/ClassWithFizzNumber;<init>
.desc Ljava/lang/Integer;
.desc ()V
.desc (Ljava/lang/Integer;Z)V

.method public static main([Ljava/lang/String;)V
getstatic java/lang/System/out Ljava/io/PrintStream;
new com/example/custom/ClassWithFizzNumber
dup
bipush 42
invokestatic java/lang/Integer/valueOf(I)Ljava/lang/Integer;
invokespecial com/example/custom/ClassWithFizzNumber/<init>(Ljava/lang/Integer;)V
invokevirtual java/io/PrintStream/println(Ljava/lang/Object;)V
return
.limit locals 1
.limit stack 4
.end method
.dep java/lang/Integer;valueOf
.dep java/io/PrintStream;println
.dep java/lang/System;out
.dep com/example/custom/ClassWithFizzNumber;<init>
.provide com/example/custom/ClassWithFizzNumber;main
.desc (Ljava/lang/Object;)V
.desc (I)Ljava/lang/Integer;
.desc Ljava/io/PrintStream;
.desc (Ljava/lang/Integer;)V

.method public toString()Ljava/lang/String;
.annotation invisible Llombok/Generated;
.end annotation
aload 0
getfield com/example/custom/ClassWithFizzNumber/number Ljava/lang/Integer;
areturn
.limit locals 1
.limit stack 1
.end method
.dep com/example/custom/ClassWithFizzNumber;number
.provide com/example/custom/ClassWithFizzNumber;toString
.desc Ljava/lang/Integer;

.method private static validate(Ljava/lang/Integer;)V
new com/example/custom/FizzBuzzNumberValidator
dup
invokespecial com/example/custom/FizzBuzzNumberValidator/<init>()V
aload 0
aconst_null
invokevirtual com/example/custom/FizzBuzzNumberValidator/isValid(Ljava/lang/Integer;Ljakarta/validation/ConstraintValidatorContext;)Z
ifne L0
new java/lang/IllegalArgumentException
dup
ldc "number must be divisible by 3 or 5"
invokespecial java/lang/IllegalArgumentException/<init>(Ljava/lang/String;)V
athrow
L0:
.stack
locals Object java/lang/Integer
.end stack
return
.limit locals 1
.limit stack 3
.end method
.dep com/example/custom/FizzBuzzNumberValidator;<init>
.dep com/example/custom/FizzBuzzNumberValidator;isValid
.dep java/lang/IllegalArgumentException;<init>
.provide com/example/custom/ClassWithFizzNumber;validate
.desc ()V
.desc (Ljava/lang/Integer;Ljakarta/validation/ConstraintValidatorContext;)Z
.desc (Ljava/lang/String;)V

.method private static validate(Ljava/lang/Integer;Z)V
new com/example/custom/FizzBuzzNumberValidator
dup
invokespecial com/example/custom/FizzBuzzNumberValidator/<init>()V
aload 0
aconst_null
invokevirtual com/example/custom/FizzBuzzNumberValidator/isValid(Ljava/lang/Integer;Ljakarta/validation/ConstraintValidatorContext;)Z
ifne L0
new java/lang/IllegalArgumentException
dup
ldc "other message"
invokespecial java/lang/IllegalArgumentException/<init>(Ljava/lang/String;)V
athrow
L0:
.stack
locals Object java/lang/Integer
locals Integer
.end stack
return
.limit locals 2
.limit stack 3
.end method
.dep com/example/custom/FizzBuzzNumberValidator;<init>
.dep com/example/custom/FizzBuzzNumberValidator;isValid
.dep java/lang/IllegalArgumentException;<init>
.provide com/example/custom/ClassWithFizzNumber;validate
.desc ()V
.desc (Ljava/lang/Integer;Ljakarta/validation/ConstraintValidatorContext;)Z
.desc (Ljava/lang/String;)V
